### PR TITLE
#27979  new lang vars endpoint

### PR DESCRIPTION
**Pull request by fabrizzio-dotCMS**
### Proposed Changes
* The main goal of this PR is to introduce the new methods that will be used to replace the old Language Keys stored as property files 
* These methods have been added to LanguageVariableAPI
* My Main concern here is that the new methods are designed to simply return the LanguageVariables no user or permission is contemplated here. I need to know if we're gonna need to do that.  
* Even though there were a few other methods here designed to return KeyValue objects these have a problem with being index dependant 
* A New interface LanguageVariable has been created. it extends KeyValue but explicitly adds the intention to have a language associated. Also, there is a new DataGen related to these classes.  
* These LanguageVariables are cached using the same LanguageCache but a separate group.  All the Language Variables are cached per language. so if we update or unpublish a LanguageVariable let's say (and cache invalidation is needed) This will not affect other LanguageVaribales of a different language living cache.
* A new Endpoint for retrieving LanguageVariables has been added.
* The respective postman test was added too.  
* Changes were made to ESContentletAPIImpl so the respective cache invalidation can be accomplished.
* Several issues were outlined on ESContentletAPIImpl by sonar so they were corrected.

### CODE REVIEW

**Comment by nollymar** [comment]
@fabrizzio-dotCMS about your concern here "My Main concern here is that the new methods are designed to simply return the LanguageVariables no user or permission is contemplated here. I need to know if we're gonna need to do that." ....we agreed that the language variables would inherit permissions from the LanguageVariable content type

**Comment by fabrizzio-dotCMS** [comment]
> @fabrizzio-dotCMS about your concern here "My Main concern here is that the new methods are designed to simply return the LanguageVariables no user or permission is contemplated here. I need to know if we're gonna need to do that." ....we agreed that the language variables would inherit permissions from the LanguageVariable content type

Yeah, we share the same concern. There are two main methods here. The one that returns Paginated LaguageVariables. That one I intend to use to populate the new screen.  The admin role guards that one. So I don't think We're gonna a user there. 

The second method I'm introducing here is the one that I intend to use to replace the properties LanguageLey on that one I introduced a User. And I perform a permission check. My reasoning here was we can always call it using system user. to get everything

**Comment by nollymar** [comment]
@wezell do we want to get language variables from the database? I understand that we don't want to depend on ES for it and @fabrizzio-dotCMS is caching those records after hitting the database. However, considering that language variables are pieces of content now, shouldn't we just leverage on the index the same way we do in Content Search for example? 

On the other hand, @fabrizzio-dotCMS has another point that is totally valid ... we shouldn't rely on ES when a user wants to use a specific language variable.

We would like your input on it.

File: dotCMS/src/main/java/com/dotcms/content/elasticsearch/business/ESContentFactoryImpl.java [hunk 1]
```diff
@@ -1155,6 +1159,142 @@ protected List<Contentlet> findByStructure(String structureInode, Date maxDate,
...
+            final int offset, final int limit,
+            final String orderBy, final boolean working) throws DotDataException, DotStateException {
+
+       final DotConnect dotConnect = new DotConnect();
+       String workingOrLiveNode = WORKING_INODE;
+       if(!working) {
+           workingOrLiveNode = LIVE_INODE;
+       }
+
+       final String orderByClause = StringUtils.isNotEmpty(orderBy) ? ORDER_BY + orderBy : StringPool.BLANK;
```
Comment (jdotcms) [inline dotCMS/src/main/java/com/dotcms/content/elasticsearch/business/ESContentFactoryImpl.java:1237]: you are concatening this value, should sanitized before concat

File: dotCMS/src/main/java/com/dotcms/languagevariable/business/LanguageVariableAPIImpl.java [hunk 2]
```diff
@@ -120,4 +146,105 @@ private String getValueFromDefaultLanguage(final String key, final User user, fi
...
+   *
+   * @param langId - The ID of the language that the variable was created for.
+   * @param user
+   * @return
+   * @throws DotDataException
+   */
+  @CloseDBIfOpened
+  @Override
+  public List<LanguageVariable> findVariables(final long langId, final User user) throws DotDataException {
+      final ContentletFactory contentletFactory = FactoryLocator.getContentletFactory();
```
Comment (jdotcms) [inline dotCMS/src/main/java/com/dotcms/languagevariable/business/LanguageVariableAPIImpl.java:177]: log inputs

File: dotCMS/src/main/java/com/dotcms/rest/AbstractPaginationContext.java [hunk 3]
```diff
@@ -0,0 +1,63 @@
+package com.dotcms.rest;
+
+public class AbstractPaginationContext {
```
Comment (jdotcms) [inline dotCMS/src/main/java/com/dotcms/rest/AbstractPaginationContext.java:3]: nice abstraction

Comment (jdotcms) [reply dotCMS/src/main/java/com/dotcms/rest/AbstractPaginationContext.java:3]: but create a buider

File: dotCMS/src/main/java/com/dotcms/rest/api/v2/languages/LanguagesResource.java [hunk 4]
```diff
@@ -410,6 +408,30 @@ public Response getAllMessages (
...
 
+    @GET
+    @Path("/variables")
+    @JSONP
+    @NoCache
+    @Produces({MediaType.APPLICATION_JSON, "application/javascript"})
+    public Response getVariables(
+            @Context final HttpServletRequest request,
+            @Context final HttpServletResponse response,
+            @BeanParam final PaginationContext paginationContext) throws DotDataException {
```
Comment (jdotcms) [inline dotCMS/src/main/java/com/dotcms/rest/api/v2/languages/LanguagesResource.java:419]: keep in mind we already have some standard to do pagination by PaginationUtil, take a look

**Comment by fabrizzio-dotCMS** [comment]
closed in favor of https://github.com/dotCMS/core/pull/28429

**Pull request closed without merging**